src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/JsonFunctionCallUtilities.cs

```diff
@@ -74,179 +103,141 @@ internal static partial class FunctionCallHelpers
-            // problematic types above. If it becomes more common that schema creation
-            // could fail expensively, we'll want to track whether inference was already
-            // attempted and avoid doing so on subsequent accesses if it was.
-            return TrueJsonSchema;
-        }
+        _ = Throw.IfNull(parameterName);
+        _ = Throw.IfNull(serializerOptions);
+        inferenceOptions ??= JsonSchemaInferenceOptions.Default;
+
+        FunctionParameterKey key = (
+            type,
+            parameterName,
+            description,
+            hasDefaultValue,
+            defaultValue,
+            IncludeSchemaUri: false,
+            inferenceOptions.DisallowAdditionalProperties,
+            inferenceOptions.IncludeTypeInEnumSchemas);
+
+        return GetJsonSchemaCached(serializerOptions, key);
     }
 
-    /// <summary>Infers a JSON schema from the return parameter.</summary>
-    /// <param name="type">The type of the return parameter.</param>
-    /// <param name="options">The options used to extract the schema from the specified type.</param>
+    /// <summary>Infers a JSON schema for the specified type.</summary>
+    /// <param name="type">The type for which to generate the schema.</param>
+    /// <param name="serializerOptions">The options used to extract the schema from the specified type.</param>
+    /// <param name="description">The description of the parameter.</param>
+    /// <param name="hasDefaultValue">Whether the parameter is optional.</param>
+    /// <param name="defaultValue">The default value of the optional parameter, if applicable.</param>
+    /// <param name="inferenceOptions">The options controlling schema inference.</param>
     /// <returns>A <see cref="JsonElement"/> representing the schema.</returns>
-    public static JsonElement InferReturnParameterJsonSchema(Type? type, JsonSerializerOptions options)
+    public static JsonElement InferJsonSchema(
+        Type? type,
+        JsonSerializerOptions serializerOptions,
+        string? description = null,
+        bool hasDefaultValue = false,
+        object? defaultValue = null,
+        JsonSchemaInferenceOptions? inferenceOptions = null)
     {
-        _ = Throw.IfNull(options);
+        _ = Throw.IfNull(serializerOptions);
+        inferenceOptions ??= JsonSchemaInferenceOptions.Default;
+
+        FunctionParameterKey key = (
+            type,
+            ParameterName: null,
+            description,
+            hasDefaultValue,
+            defaultValue,
+            inferenceOptions.IncludeSchemaKeyword,
+            inferenceOptions.DisallowAdditionalProperties,
+            inferenceOptions.IncludeTypeInEnumSchemas);
+
+        return GetJsonSchemaCached(serializerOptions, key);
+    }
 
+    private static JsonElement GetJsonSchemaCached(JsonSerializerOptions options, FunctionParameterKey key)
+    {
         options.MakeReadOnly();
```

Comment:
Do we need to document anything about using a JSO with these inference methods will make it read only? Or is that implied?

Reply:
I would say that's generally speaking implied when trying to use a JSO for anything meaningful.